mother/src/drive.c

```diff
@@ -0,0 +1,252 @@
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/can.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <app_version.h>
+#include <math.h>
+
+struct DiffDriveOdometryConfig {
+	int timestamp;
+	float x, y, heading, wheel_separation, left_wheel_radius, right_wheel_radius;
+	int velocity_rolling_window_size;
+};
+struct DiffDriveOdometry {
+	const struct DiffDriveOdometryConfig config;
+	int timestamp;
+	float x, y, heading, left_wheel_old_pos, right_wheel_old_pos, linear, angular;
+	struct FloatRollingMeanAccumulator *linear_accumulator, *angular_accumulator;
+};
+
+int diffdrive_odometry_update_from_velocity(struct DiffDriveOdometry *odom, float left_vel,
+					    float right_vel, const int time);
+
+struct DiffDriveTwist {
+	float linear_x;
+	float angular_z;
+};
+
+enum DiffDriveUpdateType {
+	POSITION_FEEDBACK,
+	VELOCITY_FEEDBACK
+};
+
+struct DiffDriveConfig {
+	float wheel_separation;
+	int wheels_per_side; // actually 2, but both are controlled by 1 signal
+	float wheel_radius;
+
+	float wheel_separation_multiplier;
+	float left_wheel_radius_multiplier;
+	float right_wheel_radius_multiplier;
+	int update_type;
+};
+
+struct DiffDrive {
+	const struct DiffDriveConfig config;
+	int previous_update_timestamp;
+	const struct DiffDriveOdometryConfig odometry_config;
+	struct DiffDriveOdometry *odometry;
+	int (*feedback_callback)(float *feedback_buffer, int buffer_len, int wheels_per_side);
+	int (*velocity_callback)(const float *velocity_buffer, int buffer_len, int wheels_per_side);
+};
+
+void *diffdrive_init(struct DiffDriveConfig *config)
+{
+	// TODO: Make a constructor here
+	return (void *)0xDEADBEEF;
+}
```

Comment:
Will you finish this? @kknives

Reply:
Finished it in https://github.com/waseemR02/kyvernitis/pull/21/commits/dc893cc8d4ddd67fd01bad10d8a408abe2e24a2e